walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,107 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+import java.util.TreeSet;
 
 public final class FindMeetingQuery {
+
+  /**
+   * Given the collection of events, this method returns a collection of all time ranges
+   * that are viable for the requested meeting.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> validTimeRanges = new ArrayList<>();
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return validTimeRanges;
+    }
+
+    Set<TimeRange> unavailableTime = determineUnavailableTime(events, request);
+    Iterator<TimeRange> iterator = unavailableTime.iterator();
+    TimeRange currentEvent = (iterator.hasNext()) ?  iterator.next() : null;
+    if (currentEvent != null) {
+      if (TimeRange.START_OF_DAY < currentEvent.start()
+          && currentEvent.start() - TimeRange.START_OF_DAY >= request.getDuration()) {
+        validTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.start(), false));
+      }
+
+      while (iterator.hasNext()) {
+        TimeRange nextEvent = iterator.next();
+
+        if (nextEvent.start() - currentEvent.end() >= request.getDuration()) {
+          validTimeRanges.add(TimeRange.fromStartEnd(currentEvent.end(), nextEvent.start(), false));
+        }
+        currentEvent = nextEvent;
+      }
+
+      if (currentEvent.end() < TimeRange.END_OF_DAY
+          && TimeRange.END_OF_DAY - currentEvent.end() + 1 >= request.getDuration()) {
+        validTimeRanges.add(TimeRange.fromStartEnd(currentEvent.end(), TimeRange.END_OF_DAY, true));
+      }
+    } else {  // There are no times taken in the day.
+      validTimeRanges.add(TimeRange.WHOLE_DAY);
+    }
+    return validTimeRanges;
   }
+
+  /**
+   * @return a Set containing all the unavailable time in the day.
+   */
+  private Set<TimeRange> determineUnavailableTime(Collection<Event> events, MeetingRequest request) {
+    Collection<String> attendees = request.getAttendees();
+    TreeSet<TimeRange> timeTaken = new TreeSet<>((range1, range2) -> {
+      if (range1.start() < range2.start()) {
+        return -1;
+      } else if (range1.start() > range2.start()) {
+        return 1;
+      } else {
+        return 0;
+      }
+    });
+
+    for (String attendee : attendees) {
+      for (Event event : events) {
+        if (event.getAttendees().contains(attendee)) {
+          timeTaken.add(event.getWhen());
+        }
+      }
+    }
```

Comment:
[Optional] Oooo, we've got a fun problem here.  The current solution is O(m*n) where m is the number of attendees and n is the sum of the number of attendees of all events.  However, if we are willing to use a bit of storage, we can solve this in O(m+n).  Can you see how?

I'm going to bring this up on Monday at the sync because I think it's a good algorithms question.  If you do come up with a better solution before then, don't feel the need to make the change in this PR.  In fact, I'd prefer you didn't and instead added a TODO which you can come back to later if you have time.

Reply:
I will add a TODO for now